Cgen error with Sumtype and Options

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.11 14ef765, press to see full `v doctor` output</summary>

|V full version      |V 0.4.11 c221b3226b50b40472864ab261e781a44bf6a5b9.14ef765
|:-------------------|:-------------------
|OS                  |windows, Microsoft Windows 11 Pro 22635 64-bit
|Processor           |8 cpus, 64bit, little endian, 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz  
|Memory              |31.88GB/63.67GB
|                    |
|V executable        |D:\software\v\v.exe
|V last modified time|2025-09-17 09:52:05
|                    |
|V home dir          |OK, value: D:\software\v
|VMODULES            |OK, value: C:\Users\Sammy\.vmodules
|VTMP                |OK, value: C:\Users\Sammy\AppData\Local\Temp\v_0
|Current working dir |OK, value: D:\Code\v\unnamed_compiler\v_bugs\cgen_error_match_true
|                    |
|Git version         |git version 2.51.0.windows.1
|V git status        |weekly.2025.16-787-g14ef7651
|.git/config present |true
|                    |
|cc version          |cc (Rev2, Built by MSYS2 project) 13.2.0
|gcc version         |gcc (Rev2, Built by MSYS2 project) 13.2.0
|clang version       |clang version 21.1.1
|msvc version        |N/A
|tcc version         |tcc version 0.9.27 (x86_64 Windows)
|tcc git status      |thirdparty-windows-amd64 b425ac82
|emcc version        |emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 4.0.3 (a9651ff57165f5710bb09a5fe52590fd6ddb72df)
|glibc version       |N/A

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  main.v && D:\Code\v\unnamed_compiler\v_bugs\cgen_error_match_true\main.exe`
```v
module main

enum Test {
	a
	b
}

type Sumtype = Test | int

fn main() {
	t := ?Sumtype(Test.a)
	match true {
		t == ?Sumtype(5) {
			println('5')
		}
		t == ?Sumtype(Test.a) {
			println('a')
		}
		else {
			println('Unknown')
		}
	}
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: d:/software/v/thirdparty/tcc/include/winapi/synchapi.h:167: warning: CreateMutex redefined
cc: C:/Users/Sammy/AppData/Local/Temp/v_0/main.01K5MC67DQVY0JCAW0JS0ND7JD.tmp.c:5082: error: cannot convert 'struct main__Sumtype' to 'struct _option_main__Sumtype'
...
cc: C:/Users/Sammy/AppData/Local/Temp/v_0/main.01K5MC67DQVY0JCAW0JS0ND7JD.tmp.c:5082: error: cannot convert 'struct main__Sumtype' to 'struct _option_main__Sumtype'
(note: the original output was 8 lines long; it was truncated to its first 2 lines + the last line)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

No Cgen error



> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/infix.v
+++ b/vlib/v/gen/c/infix.v
@@ -344,6 +344,8 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 				if node.op == .ne {
 					g.write('!')
 				}
+				tmp_left_is_opt := g.left_is_opt
+				g.left_is_opt = true
 				g.write('${ptr_typ}_sumtype_eq(')
 				if left.typ.is_ptr() {
 					g.write('*'.repeat(left.typ.nr_muls()))
@@ -355,6 +357,7 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 				}
 				g.expr(node.right)
 				g.write(')')
+				g.left_is_opt = tmp_left_is_opt
 			}
 			.interface {
 				ptr_typ := g.equality_fn(left.unaliased)

--- a/vlib/v/tests/options/option_match_eq_test.v
+++ b/vlib/v/tests/options/option_match_eq_test.v
@@ -1,0 +1,23 @@
+module main
+
+enum Test {
+	a
+	b
+}
+
+type Sumtype = Test | int
+
+fn test_main() {
+	t := ?Sumtype(Test.a)
+	match true {
+		t == ?Sumtype(5) {
+			assert false
+		}
+		t == ?Sumtype(Test.a) {
+			assert true
+		}
+		else {
+			assert false
+		}
+	}
+}

